src/main/java/com/cryptomorin/xseries/particles/ParticleDisplay.java

```diff
@@ -1794,4 +1833,221 @@ public Vector mul(Vector point) {
+
+        /**
+         * If this data doesn't support the given particle type but can be
+         * converted to a type that does, this method should return the appropriate
+         * ParticleData for that particle type.
+         * Used for converting RGB particle data to note block particle data.
+         */
+        public default ParticleData transform(XParticle particle) {
+            return this;
+        }
+    }
+
+    public static class RGBParticleColor implements ParticleData {
+        private final Color color;
+        public RGBParticleColor(Color color) {
+            this.color = color;
+        }
+
+        public RGBParticleColor(int r, int g, int b) {
+            this(new Color(r, g, b));
+        }
+
+        @Override
+        public boolean isOffsetBased(XParticle particle) {
+            // All colors were offset-based until 1.13
+            if (!ISFLAT) return true;
+            // ENTITY_EFFECT particle uses the offset fields for color on 1.20.4 and below.
+            return particle == XParticle.ENTITY_EFFECT && particle.isSupported() && particle.get().getDataType() == Void.class;
+        }
+
+        @Override
+        public double[] offsetValues() {
+            return new double[]{color.getRed() / 255d, color.getGreen() / 255d, color.getBlue() / 255d};
+        }
+
+        public Object data(XParticle particle, double extra) {
+            if (particle == XParticle.DUST) {
+                return new Particle.DustOptions(org.bukkit.Color.fromRGB(color.getRed(), color.getGreen(), color.getBlue()), (float) extra);
+            } else if (particle == XParticle.DUST_COLOR_TRANSITION) {
+                org.bukkit.Color color = org.bukkit.Color.fromRGB(this.color.getRed(), this.color.getGreen(), this.color.getBlue());
+                return new Particle.DustTransition(color, color, (float) extra);
+            }
+            if (SUPPORTS_ALPHA_COLORS) {
+                return org.bukkit.Color.fromARGB(color.getAlpha(), color.getRed(), color.getGreen(), color.getBlue());
+            } else {
+                return org.bukkit.Color.fromRGB(color.getRed(), color.getGreen(), color.getBlue());
+            }
+        }
+
+        @Override
+        public void serialize(ConfigurationSection section) {
+            StringJoiner colorJoiner = new StringJoiner(", ");
+            colorJoiner.add(Integer.toString(color.getRed()));
+            colorJoiner.add(Integer.toString(color.getGreen()));
+            colorJoiner.add(Integer.toString(color.getBlue()));
+            section.set("color", colorJoiner.toString());
+        }
+
+        @Override
+        public ParticleData transform(XParticle particle) {
+            if (particle == XParticle.NOTE) {
+                return asNoteColor();
+            }
+            return this;
+        }
+
+        public NoteParticleColor asNoteColor() {
+            return new NoteParticleColor(findNearestNoteColor(color));
+        }
```

Comment:
This small single line is used once, we can just inline it inside transform method.